moleculemaker/mmli-backend
# Feat/add novostoic service

# Problem
We want to test job creation, job status query, and job result fetch for novostoic service. However, we don't have novostoic image, and new service needs to be implemented.

# Approach
- add novostoic fake service that uses job management workflow, and inject fake data in frontend.

# How to Test
There are four services in novostoic. Each service touches four endpoints. So repeat the following procedure for each jobType: `novostoic-optstoic`, ~~`novostoic-novostoic`, `novostoic-enzrank`, `novostoic-dgpredictor`~~ (no need to repeat for others because they internally uses `novostoic-optstoic` api)

## Test Job Creation
1. navigate to `http://localhost:8080/docs#/Jobs/create_job__job_type__jobs_post`
2. click try it out
3. enter a jobType, and change the requestBody to
```
{
  "job_info": "string",
  "email": "string"
}
```

4. click execute
5. get the `job_id` field from the response

## Test Job running
6. navigate to http://localhost:8080/docs#/Novostoic
7. find the jobType of the job just created
8. click try it out
9. paste the job_id to jobId field, and no need to touch other fields
10. click execute
11. The response body should be something like this
```
{
  "jobId": "d3ed8807-e4a5-4cdb-81c5-67b1f6658acd",
  "submitted_at": "2024-04-12T23:12:29.923455"
}
```

## Test Job status
12. navigate to http://localhost:8080/docs#/Jobs/get_job_by_type_and_job_id_and_run_id__job_type__jobs__job_id___run_id__get
13. enter job_type, job_id, and a run id of 0
14. click execute
15. the response should be something like this
```
{
  "job_info": "string",
  "run_id": "0",
  "type": "novostoic-optstoic",
  "command": "",
  "time_start": 1712963549,
  "deleted": 0,
  "email": "string",
  "phase": "completed",
  "job_id": "d3ed8807-e4a5-4cdb-81c5-67b1f6658acd",
  "image": "",
  "time_created": 1712963416,
  "time_end": 1712963552,
  "user_agent": ""
}
```
a phase of completed means the job has finished successfully

## Test result fetch
16. navigate to http://localhost:8080/docs#/Files/get_results__bucket_name__results__job_id__get
17. enter bucket_name = {jobType}
18. enter job_id
19. click execute
20. the response body should be `{jobType}:pong`
